File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        res = 0
        h= {}
        maxF = 0
        l = 0
        for r in range(len(s)):
            h[s[r]] = 1 + h.get(s[r],0)
            maxF = max(h[s[r]], maxF)
            
            while r-l+1 - maxF > k:
                h[s[l]] -=1
                l+=1
            res = max(res, r-l+1)
        return res

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character,Integer> counts = new HashMap<>();
        int longest = 0;
        int l = 0;
        int maxFreq = 0;
        for(int r =0; r<s.length();r++){
            // if (counts.containsKey(s.charAt(r))){
            //     counts.put(counts.get(s.charAt(r)) +1)
            // }
            counts.put(s.charAt(r), counts.containsKey(s.charAt(r)) ? counts.get(s.charAt(r)) + 1 : 1);
            maxFreq = Math.max(maxFreq, counts.get(s.charAt(r)));
            
            while (r-l+1 - maxFreq >k){
                counts.put(s.charAt(l), counts.get(s.charAt(l)) - 1);
                l++;
            }
            longest = Math.max(longest, r-l+1);
        }
        return longest;
    }
}